src/librustc_typeck/astconv.rs

```diff
@@ -648,7 +656,87 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+                                        .iter()
+                                        .find(|param| {
+                                            if let GenericParamDefKind::Lifetime = param.kind {
+                                                true
+                                            } else {
+                                                false
+                                            }
+                                        })
+                                        .is_some();
+                                    let is_types_present = ordered_params
+                                        .iter()
+                                        .find(|param| {
+                                            if let GenericParamDefKind::Type { .. } = param.kind {
+                                                true
+                                            } else {
+                                                false
+                                            }
+                                        })
+                                        .is_some();
+                                    let is_consts_present = ordered_params
+                                        .iter()
+                                        .find(|param| {
+                                            if let GenericParamDefKind::Const = param.kind {
+                                                true
+                                            } else {
+                                                false
+                                            }
+                                        })
+                                        .is_some();
+
+                                    Self::generic_arg_mismatch_err(
+                                        tcx.sess,
+                                        arg,
+                                        kind.descr(),
+                                        Some(&format!(
+                                            "reorder the arguments: {}{}{}: `<{}>`",
+                                            if is_lifetime_present { "lifetimes" } else { "" },
+                                            if is_types_present {
+                                                if is_lifetime_present {
+                                                    ", then types"
+                                                } else {
+                                                    "types"
+                                                }
+                                            } else {
+                                                ""
+                                            },
+                                            if is_consts_present
+                                                && tcx.sess.features_untracked().const_generics
+                                            {
+                                                if is_lifetime_present || is_types_present {
```

Comment:
This should always be true, but I suppose it's safe to leave in for future compatibility with adding in another generic parameter type.